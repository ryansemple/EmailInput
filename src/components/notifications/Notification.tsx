import React from "react";
import clsx from "clsx";

interface NotificationProps {
	text: string,
	className: string
}

/**
 * Renders a notification.
 */
const Notification = (props: NotificationProps) => (
	<div 
		className={clsx("Notifications_Notification", props.className)} 
		role="alert">
		<p>{props.text}</p>
	</div>
)

export default Notification;