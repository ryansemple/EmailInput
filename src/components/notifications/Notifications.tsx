import React, { useEffect } from "react";
import ErrorNotification from "./ErrorNotification";
import SuccessNotification from "./SuccessNotification";
import { 
	NotificationType,
	NotificationInstance
} from "../../types/Notification";
import usePrevious from "../../hooks/usePrevious";
import { 
	returnItemsInArrayNotInAnotherArrayById
} from "../../utility/Array";

interface NotificationsProps {
	notifications: NotificationInstance[],
	setNotifications: React.Dispatch<React.SetStateAction<NotificationInstance[]>>
}

const notificationTimeout: number = 6000;

const Notifications = (props: NotificationsProps) => {

	const previousProps = usePrevious({notifications: props.notifications});
	const previousNotifications: NotificationInstance[] | undefined = previousProps?.notifications;

	const destroyNotificationAfterTimeoutEnds = (noticationToBeDestroyed: NotificationInstance) => 
	{
		setTimeout(() => 
		{
			props.setNotifications(notifications =>	{
				return notifications
				.filter((notificationInstance: NotificationInstance) => 
						notificationInstance.id !== noticationToBeDestroyed.id
				);
			});
		}, notificationTimeout);
	};

	useEffect(() =>
	{	
		const notificationPropsHaveUpdated = previousNotifications?.length !==
		props.notifications.length;
		
		if (notificationPropsHaveUpdated) 
		{
			let newNotifications: NotificationInstance[] = [];

			if(!previousNotifications || previousNotifications.length === 0)
			{
				newNotifications = props.notifications;
			} 
			else 
			{
				newNotifications = returnItemsInArrayNotInAnotherArrayById
				(
					props.notifications,
					previousNotifications!
				);
			}

			for (let i: number = 0; i < newNotifications.length; i++)
			{
				destroyNotificationAfterTimeoutEnds(newNotifications[i]);
			}
		}
	}, 
	[
		props.notifications, 
		previousNotifications, 
		destroyNotificationAfterTimeoutEnds
	]);

	return (
		<div 
			className="Notifications absolute flex horizontal_center_flex full_width">
			{props.notifications
			.map((notificationInstance: NotificationInstance) => 
			{
				switch(notificationInstance.notificationType)
				{
					case NotificationType.Error:
						return (
							<ErrorNotification 
								text={notificationInstance.text}
								key={notificationInstance.id}
							/>
						)
					case NotificationType.Success:
						return (
							<SuccessNotification 
								text={notificationInstance.text}
								key={notificationInstance.id}
							/>
						)
				}
			})}
		</div>
	)
}

export default Notifications;