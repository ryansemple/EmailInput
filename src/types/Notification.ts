import ObjectWithId from "../types/ObjectWithId";

export enum NotificationType {
	Success,
	Error
}

export interface NotificationInstance extends ObjectWithId {
	text: string,
	id: string,
	notificationType: NotificationType
}