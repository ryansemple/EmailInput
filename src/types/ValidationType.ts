export enum ValidationType {
	Success,
	Error
}