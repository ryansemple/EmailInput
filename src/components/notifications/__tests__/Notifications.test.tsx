import React from "react";
import { render } from "@testing-library/react";
import Notifications from "../Notifications";
import Notification from "../../../types/Notification";
import ValidationType from "../../../types/ValidationType";

test("Renders Notifications component correctly.", () => {

	const notifications: Notification[] = [
		new Notification("success", ValidationType.Success),
		new Notification("error", ValidationType.Error)
	]

	const { container } = render(
		<Notifications
			notifications={notifications}
			setNotifications={() => {}}
		/>
	);
	expect(container).toMatchSnapshot();
});