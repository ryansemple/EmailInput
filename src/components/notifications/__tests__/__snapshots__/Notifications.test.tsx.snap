// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Notifications component correctly. 1`] = `
<div>
  <div
    class="Notifications absolute flex horizontal_center_flex full_width"
  >
    <div
      class="Notifications_Notification Notifications_Notification-Success"
      role="alert"
    >
      <p>
        success
      </p>
    </div>
    <div
      class="Notifications_Notification Notifications_Notification-Error"
      role="alert"
    >
      <p>
        error
      </p>
    </div>
  </div>
</div>
`;
